File: smsly-core/smsly_core/metrics.py
"""
Prometheus Metrics
===================
Common metrics for SMSLY microservices.
"""

from typing import Dict, Optional
from dataclasses import dataclass, field
from time import time
import structlog

logger = structlog.get_logger(__name__)


@dataclass
class MetricLabels:
    """Common labels for metrics."""
    service: str
    environment: str = "production"
    version: str = "1.0.0"


class SimpleMetrics:
    """
    Simple in-memory metrics collector.
    
    For production, use prometheus_client with pushgateway.
    """
    
    def __init__(self, labels: Optional[MetricLabels] = None):
        self.labels = labels or MetricLabels(service="unknown")
        self._counters: Dict[str, int] = {}
        self._gauges: Dict[str, float] = {}
        self._histograms: Dict[str, list] = {}
    
    def increment(self, name: str, value: int = 1, labels: Optional[Dict] = None) -> None:
        """Increment a counter."""
        key = self._make_key(name, labels)
        self._counters[key] = self._counters.get(key, 0) + value
    
    def set_gauge(self, name: str, value: float, labels: Optional[Dict] = None) -> None:
        """Set a gauge value."""
        key = self._make_key(name, labels)
        self._gauges[key] = value
    
    def observe(self, name: str, value: float, labels: Optional[Dict] = None) -> None:
        """Record a histogram observation."""
        key = self._make_key(name, labels)
        if key not in self._histograms:
            self._histograms[key] = []
        self._histograms[key].append(value)
    
    def _make_key(self, name: str, labels: Optional[Dict] = None) -> str:
        """Create a unique key for a metric."""
        if labels:
            label_str = ",".join(f"{k}={v}" for k, v in sorted(labels.items()))
            return f"{name}{{{label_str}}}"
        return name
    
    def get_counter(self, name: str, labels: Optional[Dict] = None) -> int:
        """Get counter value."""
        return self._counters.get(self._make_key(name, labels), 0)
    
    def get_histogram_stats(self, name: str, labels: Optional[Dict] = None) -> Dict:
        """Get histogram statistics."""
        key = self._make_key(name, labels)
        values = self._histograms.get(key, [])
        
        if not values:
            return {"count": 0, "sum": 0, "avg": 0, "p50": 0, "p95": 0, "p99": 0}
        
        sorted_values = sorted(values)
        count = len(values)
        
        return {
            "count": count,
            "sum": sum(values),
            "avg": sum(values) / count,
            "p50": self._percentile(sorted_values, 50),
            "p95": self._percentile(sorted_values, 95),
            "p99": self._percentile(sorted_values, 99),
        }
    
    def _percentile(self, sorted_values: list, percentile: int) -> float:
        """Calculate percentile from sorted values."""
        if not sorted_values:
            return 0
        idx = int(len(sorted_values) * percentile / 100)
        return sorted_values[min(idx, len(sorted_values) - 1)]
    
    def export_prometheus(self) -> str:
        """Export metrics in Prometheus text format."""
        lines = []
        
        # Labels
        base_labels = f'service="{self.labels.service}",env="{self.labels.environment}"'
        
        # Counters
        for key, value in self._counters.items():
            name = key.split('{')[0]
            extra_labels = key[len(name):] if '{' in key else ''
            full_labels = f'{{{base_labels}{"," + extra_labels[1:-1] if extra_labels else ""}}}'
            lines.append(f'{name}_total{full_labels} {value}')
        
        # Gauges
        for key, value in self._gauges.items():
            name = key.split('{')[0]
            extra_labels = key[len(name):] if '{' in key else ''
            full_labels = f'{{{base_labels}{"," + extra_labels[1:-1] if extra_labels else ""}}}'
            lines.append(f'{name}{full_labels} {value}')
        
        # Histograms (simplified as summary)
        for key in self._histograms:
            stats = self.get_histogram_stats(key.split('{')[0])
            name = key.split('{')[0]
            lines.append(f'{name}_count{{{base_labels}}} {stats["count"]}')
            lines.append(f'{name}_sum{{{base_labels}}} {stats["sum"]}')
        
        return '\n'.join(lines)


class Timer:
    """Context manager for timing operations."""
    
    def __init__(self, metrics: SimpleMetrics, name: str, labels: Optional[Dict] = None):
        self.metrics = metrics
        self.name = name
        self.labels = labels
        self._start: Optional[float] = None
    
    def __enter__(self):
        self._start = time()
        return self
    
    def __exit__(self, *args):
        if self._start:
            duration = time() - self._start
            self.metrics.observe(self.name, duration, self.labels)


# Pre-defined metric names
class MetricNames:
    # Request metrics
    HTTP_REQUESTS_TOTAL = "http_requests"
    HTTP_REQUEST_DURATION = "http_request_duration_seconds"
    HTTP_REQUEST_SIZE = "http_request_size_bytes"
    HTTP_RESPONSE_SIZE = "http_response_size_bytes"
    
    # Messaging metrics
    MESSAGES_SENT_TOTAL = "smsly_messages_sent"
    MESSAGES_DELIVERED_TOTAL = "smsly_messages_delivered"
    MESSAGES_FAILED_TOTAL = "smsly_messages_failed"
    MESSAGE_SEND_DURATION = "smsly_message_send_duration_seconds"
    
    # Queue metrics
    QUEUE_DEPTH = "smsly_queue_depth"
    QUEUE_PROCESSING_TIME = "smsly_queue_processing_seconds"
    
    # Database metrics
    DB_CONNECTIONS_ACTIVE = "smsly_db_connections_active"
    DB_QUERY_DURATION = "smsly_db_query_duration_seconds"
    
    # Rate limiting metrics
    RATE_LIMIT_HITS = "smsly_rate_limit_hits"
    RATE_LIMIT_ALLOWED = "smsly_rate_limit_allowed"

File: smsly-rust/smsly-core/src/metrics.rs
use lazy_static::lazy_static;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::SystemTime;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MetricLabels {
    pub service: String,
    pub environment: String,
    pub version: String,
}

impl Default for MetricLabels {
    fn default() -> Self {
        Self {
            service: "unknown".to_string(),
            environment: "production".to_string(),
            version: "1.0.0".to_string(),
        }
    }
}

pub struct SimpleMetrics {
    labels: MetricLabels,
    counters: Mutex<HashMap<String, i64>>,
    gauges: Mutex<HashMap<String, f64>>,
    histograms: Mutex<HashMap<String, Vec<f64>>>,
}

impl SimpleMetrics {
    pub fn new(labels: Option<MetricLabels>) -> Self {
        Self {
            labels: labels.unwrap_or_default(),
            counters: Mutex::new(HashMap::new()),
            gauges: Mutex::new(HashMap::new()),
            histograms: Mutex::new(HashMap::new()),
        }
    }

    fn make_key(&self, name: &str, labels: &Option<HashMap<String, String>>) -> String {
        if let Some(l) = labels {
            let mut sorted_labels: Vec<_> = l.iter().collect();
            sorted_labels.sort_by_key(|a| a.0);
            let label_str: Vec<String> = sorted_labels
                .iter()
                .map(|(k, v)| format!("{}={}", k, v))
                .collect();
            format!("{}{{{}}}", name, label_str.join(","))
        } else {
            name.to_string()
        }
    }

    pub fn increment(&self, name: &str, value: i64, labels: Option<HashMap<String, String>>) {
        let key = self.make_key(name, &labels);
        let mut counters = self.counters.lock().unwrap();
        *counters.entry(key).or_insert(0) += value;
    }

    pub fn set_gauge(&self, name: &str, value: f64, labels: Option<HashMap<String, String>>) {
        let key = self.make_key(name, &labels);
        let mut gauges = self.gauges.lock().unwrap();
        gauges.insert(key, value);
    }

    pub fn observe(&self, name: &str, value: f64, labels: Option<HashMap<String, String>>) {
        let key = self.make_key(name, &labels);
        let mut histograms = self.histograms.lock().unwrap();
        histograms.entry(key).or_default().push(value);
    }

    fn percentile(sorted_values: &[f64], percentile: f64) -> f64 {
        if sorted_values.is_empty() {
            return 0.0;
        }
        let idx = (sorted_values.len() as f64 * percentile / 100.0) as usize;
        sorted_values[std::cmp::min(idx, sorted_values.len() - 1)]
    }

    pub fn get_histogram_stats(
        &self,
        name: &str,
        labels: Option<HashMap<String, String>>,
    ) -> HashMap<String, f64> {
        let key = self.make_key(name, &labels);
        let histograms = self.histograms.lock().unwrap();

        let mut stats = HashMap::new();
        if let Some(values) = histograms.get(&key) {
            let mut sorted_values = values.clone();
            sorted_values.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
            let count = sorted_values.len() as f64;
            let sum: f64 = sorted_values.iter().sum();

            stats.insert("count".to_string(), count);
            stats.insert("sum".to_string(), sum);
            stats.insert(
                "avg".to_string(),
                if count > 0.0 { sum / count } else { 0.0 },
            );
            stats.insert("p50".to_string(), Self::percentile(&sorted_values, 50.0));
            stats.insert("p95".to_string(), Self::percentile(&sorted_values, 95.0));
            stats.insert("p99".to_string(), Self::percentile(&sorted_values, 99.0));
        } else {
            stats.insert("count".to_string(), 0.0);
            stats.insert("sum".to_string(), 0.0);
            stats.insert("avg".to_string(), 0.0);
            stats.insert("p50".to_string(), 0.0);
            stats.insert("p95".to_string(), 0.0);
            stats.insert("p99".to_string(), 0.0);
        }
        stats
    }
}

lazy_static! {
    pub static ref GLOBAL_METRICS: SimpleMetrics = SimpleMetrics::new(None);
}

pub fn track_metric(name: &str, metadata: HashMap<String, serde_json::Value>) {
    let mut labels = HashMap::new();
    let mut val = 1.0;

    for (k, v) in metadata {
        if let Some(n) = v.as_f64() {
            if k == "duration_ms" || k == "latency" || k == "value" {
                val = n;
            }
        }
        if let Some(s) = v.as_str() {
            labels.insert(k, s.to_string());
        } else {
            labels.insert(k, v.to_string());
        }
    }

    GLOBAL_METRICS.observe(name, val, Some(labels));
}

pub struct Timer<'a> {
    metrics: &'a SimpleMetrics,
    name: String,
    labels: Option<HashMap<String, String>>,
    start: Option<SystemTime>,
}

impl<'a> Timer<'a> {
    pub fn new(
        metrics: &'a SimpleMetrics,
        name: &str,
        labels: Option<HashMap<String, String>>,
    ) -> Self {
        Self {
            metrics,
            name: name.to_string(),
            labels,
            start: None,
        }
    }

    pub fn start(&mut self) {
        self.start = Some(SystemTime::now());
    }

    pub fn stop(&self) {
        if let Some(start) = self.start {
            let duration = start.elapsed().unwrap_or_default().as_secs_f64();
            self.metrics
                .observe(&self.name, duration, self.labels.clone());
        }
    }
}

pub struct MetricNames;

impl MetricNames {
    pub const HTTP_REQUESTS_TOTAL: &'static str = "http_requests";
    pub const HTTP_REQUEST_DURATION: &'static str = "http_request_duration_seconds";
    pub const MESSAGES_SENT_TOTAL: &'static str = "smsly_messages_sent";
}
